src/data/CCBV.ml

```diff
@@ -1,80 +1,117 @@
+let __count_bits n =
+  let rec recurse count n =
+    if n = 0 then count else recurse (count+1) (n land (n-1))
+  in
+  recurse 0 n
 
-(* only ones *)
-let __all_ones = __shift 0 __width
+(*  Can I access the "private" members in testing? $Q
+  (Q.int_bound (Sys.word_size - 1)) (fun i -> __count_bits __lsb_masks.(i) = i)
+*)
 
 type t = {
   mutable a : int array;
+  mutable size : int;
 }
 
-let empty () = { a = [| |] }
+let length t = t.size
+
+let empty () = { a = [| |] ; size = 0 }
+
+let __to_array_legnth size =
+  if size mod __width = 0 then size / __width else (size / __width) + 1
 
 let create ~size default =
-  if size = 0 then { a = [| |] }
+  if size = 0 then { a = [| |] ; size }
   else begin
-    let n = if size mod __width = 0 then size / __width else (size / __width) + 1 in
+    let n = __to_array_legnth size in
     let arr = if default
       then Array.make n __all_ones
       else Array.make n 0
     in
     (* adjust last bits *)
-    if default && (size mod __width) <> 0
-    then arr.(n-1) <- __shift 0 (size - (n-1) * __width);
-    { a = arr }
+    let r = size mod __width in
+    if default && r <> 0
+    then Array.unsafe_set arr (n-1) __lsb_masks.(r);
+    { a = arr; size }
   end
 
+(*$Q
+  (Q.pair Q.small_int Q.bool) (fun (size, b) -> create ~size b |> length = size)
+*)
+
 (*$T
   create ~size:17 true |> cardinal = 17
-  create ~size:32 true |> cardinal= 32
+  create ~size:32 true |> cardinal = 32
   create ~size:132 true |> cardinal = 132
   create ~size:200 false |> cardinal = 0
   create ~size:29 true |> to_sorted_list = CCList.range 0 28
 *)
 
-let copy bv = { a=Array.copy bv.a; }
+let copy bv = { a = Array.copy bv.a ; size = bv.size }
 
 (*$Q
   (Q.list Q.small_int) (fun l -> \
     let bv = of_list l in to_list bv = to_list (copy bv))
 *)
 
-let length bv = Array.length bv.a
-
-let resize bv len =
-  if len > Array.length bv.a
-  then begin
-    let a' = Array.make len 0 in
-    Array.blit bv.a 0 a' 0 (Array.length bv.a);
-    bv.a <- a'
-  end
-
-(* count the 1 bits in [n]. See https://en.wikipedia.org/wiki/Hamming_weight *)
-let __count_bits n =
-  let rec recurse count n =
-    if n = 0 then count else recurse (count+1) (n land (n-1))
-  in
-  if n < 0
-  then recurse 1 (n lsr 1)   (* only on unsigned *)
-  else recurse 0 n
+let capacity bv = __width * Array.length bv.a
 
 let cardinal bv =
   let n = ref 0 in
-  for i = 0 to length bv - 1 do
+  for i = 0 to Array.length bv.a - 1 do
     n := !n + __count_bits bv.a.(i)
   done;
   !n
 
+(*$Q
+  Q.small_int (fun size -> create ~size true |> cardinal = size)
+*)
+
+let __really_resize bv ~desired ~current size =
+  let a' = Array.make desired 0 in
+  Array.blit bv.a 0 a' 0 current;
+  bv.a <- a';
+  bv.size <- size
+
+let __grow bv size =
+  if size <= capacity bv                                   (* within capacity *)
+  then bv.size <- size
+  else                                                     (* beyond capacity *)
```

Comment:
stylistic nitpick: I tend to wrap such a `else` into `( … )`, to prevent any issue related to dangling else…
sth like
```ocaml
else (
   let foo = bar in
   let foo2 = bar2 in
   quux
)
```